src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi

```diff
@@ -94,28 +93,19 @@ cdef class CompletionQueue:
   def poll(self, deadline=None):
     return self._interpret_event(_next(self.c_completion_queue, deadline))
 
+  # All outstanding calls must be cancelled before invoking this method.
   def shutdown(self):
+    cdef gpr_timespec c_deadline
+    c_deadline = gpr_inf_future(GPR_CLOCK_REALTIME)
     with nogil:
       grpc_completion_queue_shutdown(self.c_completion_queue)
-    self.is_shutting_down = True
-
-  def clear(self):
-    if not self.is_shutting_down:
-      raise ValueError('queue must be shutting down to be cleared')
-    while self.poll().type != GRPC_QUEUE_SHUTDOWN:
-      pass
+    while not self.is_shutdown:
+      event = grpc_completion_queue_next(
+          self.c_completion_queue, c_deadline, NULL)
+      self._interpret_event(event)
+    grpc_completion_queue_destroy(self.c_completion_queue)
+    self.c_completion_queue = NULL
 
   def __dealloc__(self):
-    cdef gpr_timespec c_deadline
-    c_deadline = gpr_inf_future(GPR_CLOCK_REALTIME)
-    if self.c_completion_queue != NULL:
-      # Ensure shutdown
-      if not self.is_shutting_down:
-        grpc_completion_queue_shutdown(self.c_completion_queue)
-      # Pump the queue (All outstanding calls should have been cancelled)
-      while not self.is_shutdown:
-        event = grpc_completion_queue_next(
-            self.c_completion_queue, c_deadline, NULL)
-        self._interpret_event(event)
-      grpc_completion_queue_destroy(self.c_completion_queue)
-    grpc_shutdown()
+    if self.c_completion_queue == NULL:
+      grpc_shutdown()
```

Comment:
The core objects won't react to deallocation at all. There is no way to force Cython garbage collection (before or after this PR).

Previously the cython object had to be deallocated to trigger the core object to be deallocated: but this can require first shutting down the core objects, which is unsafe to do in a `__dealloc__` method. This change decouples the cython level deallocation from the core shutdown - but does indeed require an explicit shutdown/close at the Python layer to trigger the core deallocation, which appears to be the best we can do.